### PR TITLE
Add ManageGoRules option to kazel

### PR DESCRIPTION
**Pull request by ixdy**
This should allow us to use `gazelle` along with the sourcerer and generator functionality we need from `kazel`. Part of https://github.com/kubernetes/kubernetes/issues/47558.

It is a bit messy. At some point we should probably refactor this to remove all of the go-rule functionality that we no longer need.

I've tested this on `repo-infra`, where it basically seems to work. I'll try a larger test on `test-infra` next.

/assign @mikedanese @spxtr

### CODE REVIEW

**Comment by ixdy** [comment]
seems to play nicely with gazelle in `test-infra`.

**Comment by ixdy** [comment]
hm, though it seems like the sourcerer module is not working properly if `SkipGoRules` is enabled. investigating...

**Comment by ixdy** [comment]
oh, I unintentionally made it ignore all `automanaged` rules.

we probably want to remove the `automanaged` tag from all the go rules when we switch to gazelle. I'm not sure of the best way to do that, though.

**Comment by ixdy** [comment]
to answer my last question, https://github.com/bazelbuild/buildtools/tree/master/buildozer looks promising.

**Comment by ixdy** [comment]
OK, I think I have this working now; I can run `gazelle && kazel` and things basically work as expected.

The main remaining issue is that when `gazelle` generates a new `BUILD` file, it doesn't add the `package(default_visibility = ["//visibility:public"])` and `licenses(["notice"])` lines. (It prefers adding visibility to rules directly, and I guess licenses might be out of scope for the tool, though we could ask.) `kazel` won't add them either, because the `BUILD` file already exists.

I could fix this, but I'm not sure whether this is the right place, or whether it should go in `gazelle` or if we should process through `buildozer` to add the lines.

**Comment by ixdy** [comment]
OTOH, Bazel barely uses `licenses()` in its BUILD files, and `rules_go` doesn't seem to use it anywhere, so maybe we don't need it?

**Comment by ixdy** [comment]
per some offline discussions, we probably shouldn't be adding the `licenses()` rules anyway - they're basically only used in Blaze, and the ones we're autogenerating may not even be correct.

Also, Gazelle's per-rule visibility is probably better than using default visibility, so we may want to do a cleanup there, too.

tl;dr I believe this is ready to be reviewed as-is.

**Comment by spxtr** [comment]
Does this still play properly with the openapi generated rule?

**Comment by ixdy** [comment]
Apparently it does not.
I don't entirely understand how it works, so I guess I'll dig in...

**Comment by ixdy** [comment]
OK, this seems to work now. We need to still `walkVendor` and `walkRepo` as otherwise the rules don't get populated for `https://github.com/kubernetes/repo-infra/blob/f521b5d472e00e05da5394994942064510a6e8bf/kazel/generator.go#L109`.

(At some point we should probably fix this so it doesn't depend on `walkRepo` having been run. Maybe when we strip out all of the go logic for good?)

I also inverted the boolean, so now by default `kazel` will *not* manage go rules.

**Comment by spxtr** [comment]
/lgtm
Although I'm not a fan of flipping the boolean, I don't think it's a big deal.

**Comment by ixdy** [comment]
yeah, I could go either way. I figure we're going to remove go rules in the long run, so it's really just an escape hatch in the short term if things are really broken.

**Comment by ixdy** [comment]
also, the other boolean options are positive (AddSourcesRules, VendorMultipleBuildFiles, K8sOpenAPIGen) so it felt weird to have a negative here.
